File: app/queries/v3/fetch_vod_channels_query.rb
# frozen_string_literal: true

module V3
  class FetchVodChannelsQuery < V3::ApplicationQuery
    def call
      build_object(execute(query_string))
    end

    private

    def build_object(result)
      nodes = result.dig(:data, :channels, :nodes)
      nodes.map { |node| ChannelStruct.new(node.slice(*ChannelStruct.attribute_names)) }
    end

      def query_string
      <<~GRAPHQL
        query {
          channels(isVod: true) {
            nodes {
              annictId
              name
            }
          }
        }
      GRAPHQL
    end
  end
end

File: app/frontend/javascripts/v3/queries/FetchVodChannelsQuery.ts
import gql from 'graphql-tag'

import client from '../client'
import { Channel } from '../models'
import { ApplicationQuery } from './ApplicationQuery'

const query = gql`
  query {
    channels(isVod: true) {
      nodes {
        annictId
        name
      }
    }
  }
`

export class FetchVodChannelsQuery extends ApplicationQuery {
  public async execute(): Promise<[Channel]> {
    const result = await client.query({ query: query })
    return result.data.channels.nodes.map((node): Channel => {
      return new Channel(node)
    })
  }
}
